File: python/lca.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param {TreeNode} root
    # @param {TreeNode} p
    # @param {TreeNode} q
    # @return {TreeNode}
    def lowestCommonAncestor(self, root, p, q):
        if root == None:
                return None
        if root == p or root == q:
                return root
        left = self.lowestCommonAncestor(root.left, p,q) #look for the p or q in the left subtree
        right = self.lowestCommonAncestor(root.right, p,q) #look for the p or q in the right subtree
        if left is not None and right is not None:
                return root
        if left is not None:
                return left
        else:
                return right

File: cpp/lca.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if(!root)
            return NULL;
        if(root == p || root == q)
            return root;
        if(root->left == NULL && root->right == NULL)
            return NULL;
        TreeNode* tmp = lowestCommonAncestor(root->left, p, q);
        TreeNode* tmp2 = lowestCommonAncestor(root->right, p, q);
        if(tmp && tmp2)
            return root;
        else if(tmp)
            return tmp;
        else if(tmp2)
            return tmp2;
    }
};